pip/index.py

```diff
@@ -779,8 +791,8 @@ def _link_package_versions(self, link, search_name):
                     link, 'Python version is incorrect')
                 return
         logger.debug('Found link %s, version: %s', link, version)
-
-        return InstallationCandidate(search_name, version, link)
+        link.canonical_name = search.canonical
```

Comment:
I'm not a fan either. What you propose won't work since we don't know the actual requested name in InstallRequirement all of the time. Perhaps if we say 'if we don't know it, then we can't cache it', will be sufficient.